File: adt2unit/pas/reset_player.pas
// This file is part of Adlib Tracker II (AT2).
//
// SPDX-FileType: SOURCE
// SPDX-FileCopyrightText: 2014-2024 The Adlib Tracker 2 Authors
// SPDX-License-Identifier: GPL-3.0-or-later

procedure reset_player;

var
  temp: Byte;

begin
  _dbg_enter ({$I %FILE%}, 'reset_player');

  opl2out($01,0);

  For temp := 1 to 18 do opl2out($0b0+_chan_n[temp],0);
  For temp := $080 to $08d do opl2out(temp,BYTE_NULL);
  For temp := $090 to $095 do opl2out(temp,BYTE_NULL);

  misc_register := tremolo_depth SHL 7+
                   vibrato_depth SHL 6+
                   BYTE(percussion_mode) SHL 5;

  opl2out($01,$20);
  opl2out($08,$40);
  opl3exp($0105);
  opl3exp($04+songdata.flag_4op SHL 8);

  key_off(17);
  key_off(18);
  opl2out(_instr[11],misc_register);
  For temp := 1 to 20 do reset_chan_data(temp);

  _dbg_leave; //EXIT //reset_player
end;

File: adt2unit/reset_player.c
// This file is part of Adlib Tracker II (AT2).
//
// SPDX-FileType: SOURCE
// SPDX-FileCopyrightText: 2014-2024 The Adlib Tracker 2 Authors
// SPDX-License-Identifier: GPL-3.0-or-later

void reset_player (void)
{
  DBG_ENTER ("reset_player");

  opl2out (0x01, 0);

  for (uint8_t chan = 0; chan <= 17; chan++)
    opl2out (0xB0 + _chan_n[chan], 0);

  for (uint8_t reg = 0x80; reg <= 0x8D; reg++)
    opl2out (reg, BYTE_NULL);

  for (uint8_t reg = 0x90; reg <= 0x95; reg++)
    opl2out (reg, BYTE_NULL);

  misc_register =   (tremolo_depth << 7)
                  + (vibrato_depth << 6)
                  + (percussion_mode ? 1 << 5 : 0);

  opl2out (0x01, 0x20);
  opl2out (0x08, 0x40);
  opl3exp (0x0105);
  opl3exp (0x04 + (songdata.flag_4op << 8));

  key_off (16 + 1);
  key_off (17 + 1);
  opl2out (_instr[11], misc_register);

  for (uint8_t chan = 0; chan <= 19; chan++)
    reset_chan_data (chan + 1);

  DBG_LEAVE (); //EXIT //reset_player
}
